user_tools/src/spark_rapids_tools/tools/qualx/hash_util.py

```diff
@@ -189,16 +206,48 @@ def normalize_path(node):
                 elif path_match(node, ['UnionWithLocalData', 'Union']):
                     # UnionWithLocalData/Union/X -> X
                     node = normalize_path(node['children'][0]['children'][0])
-                elif len(node['children']) == 1 and node['children'][0]['nodeName'] == 'SubqueryAdaptiveBroadcast':
-                    # X/SubqueryAdaptiveBroadcast -> X
-                    node['children'] = []
                 else:
                     # continue normalizing children
                     node['children'] = [normalize_path(child) for child in node['children']]
         return node
 
+    def normalize_broadcast_hash_join(node):
+        """Normalize a BroadcastHashJoin subtree."""
+        if isinstance(node, dict):
+            if node['nodeName'] == 'BroadcastHashJoin':
+                # remove any Project/BroadcastHashJoin/BroadcastExchange/SortMergeJoin nodes
+                remove_list = ['Project', 'BroadcastHashJoin', 'BroadcastExchange', 'SortMergeJoin']
+                children = []
+                for child in node['children']:
+                    children.extend(remove_nodes(child, remove_list))
+                node['children'] = children
+            else:
+                # continue normalizing children
+                children = [normalize_broadcast_hash_join(child) for child in node['children']]
+                node['children'] = children
+        return node
+
+    def normalize_sort_merge_join(node):
+        """Normalize a SortMergeJoin/Project/BroadcastHashJoin subtree."""
+        if isinstance(node, dict):
+            if path_match(node, ['SortMergeJoin', 'Project', 'BroadcastHashJoin']):
+                # remove Project/BroadcastHashJoin nodes
+                remove_list = ['Project', 'BroadcastHashJoin']
+                children = []
+                for child in node['children']:
+                    children.extend(remove_nodes(child, remove_list))
+                node['children'] = children
+                node['nodeName'] = 'BroadcastHashJoin'
+            else:
+                # continue normalizing children
+                node['children'] = [normalize_sort_merge_join(child) for child in node['children']]
+        return node
+
     normalized_plan = normalize_node(plan)
     normalized_plan = normalize_path(normalized_plan)
+    normalized_plan = normalize_broadcast_hash_join(normalized_plan)
+    normalized_plan = normalize_sort_merge_join(normalized_plan)
```

Comment:
It seems that we will need to refactor the code here at some point.
We probably need to chain the the normalization methods and pass the node to that chain.
Ultimately, users can then register the normalization method and its key trigger without having to hard code the method calls.

Reply:
Yeah, I had considered that, since making multiple passes through the tree seems inefficient.  Unfortunately, each pass is essentially dependent on the prior passes, e.g. we can only normalize paths after the nodes have been normalized, etc.  That said, I agree that we can likely improve/refactor this in the future.